### PR TITLE
DEV-218 Authentication data matching for registration

### PR DESCRIPTION
**Pull request by moseshll**
- Should look at this with MS in a test environment.

### CODE REVIEW

File: app/presenters/ht_registration_presenter.rb [hunk 1]
```diff
@@ -19,14 +19,18 @@ class HTRegistrationPresenter < ApplicationPresenter
   READ_ONLY_FIELDS = %i[sent received finished ip_address env].freeze
   JIRA_BASE_URL = URI.join(Otis.config.jira.site, "/jira/", "browse/").to_s.freeze
   FIELD_SIZE = 45
+  VALID_AFFILIATIONS = %r{^(faculty|staff|member|employee)}
+  AUTH_EXCEPTIONS = ["nypl"]
```
Comment (aelkiss) [inline app/presenters/ht_registration_presenter.rb:23]: Consider moving these to configuration

File: app/presenters/ht_registration_presenter.rb [hunk 2]
```diff
@@ -79,11 +83,17 @@ def show_detail_display_name
   end
 
   def show_detail_edu_person_principal_name
-    env["HTTP_X_SHIB_EDUPERSONPRINCIPALNAME"]
+    name = env["HTTP_X_SHIB_EDUPERSONPRINCIPALNAME"] || ""
+    # Check if principal name or email matches the given DSP email
+    badge = name == applicant_email ? :ok : :mismatch
```
Comment (aelkiss) [inline app/presenters/ht_registration_presenter.rb:88]: I don't know if we necessarily want to convey "danger" here -- they aren't necessarily expected to match, but it's worth checking that it at least makes sense. It's also not unexpected for this header to be missing (not all identity providers provide it)

File: app/presenters/ht_registration_presenter.rb [hunk 3]
```diff
@@ -79,11 +83,17 @@ def show_detail_display_name
...
+    name + " " + BADGES[badge].label_span
   end
 
   def show_detail_email
-    env["HTTP_X_SHIB_MAIL"]
+    email = env["HTTP_X_SHIB_MAIL"] || ""
+    # Check if principal name or email matches the given DSP email
+    badge = email == applicant_email ? :ok : :mismatch
```
Comment (aelkiss) [inline app/presenters/ht_registration_presenter.rb:95]: As with eppn - it's not "danger do not approve" if they don't match, but should invite extra scrutiny; likewise not all IdPs will send it, so not really a problem if it's missing entirely

File: app/presenters/ht_registration_presenter.rb [hunk 4]
```diff
@@ -94,8 +104,15 @@ def show_detail_geoip
...
 
   def show_detail_identity_provider
-    entity = env["HTTP_X_SHIB_IDENTITY_PROVIDER"]
-    ERB::Util.html_escape(HTInstitution.where(entityID: entity).first&.name || entity)
+    idp = env["HTTP_X_SHIB_IDENTITY_PROVIDER"] || ""
+    # Check if identity provider matches the institution picked on the signup form
+    inst = HTInstitution.where(entityID: idp).first
+    badge = ""
+    unless self.class::AUTH_EXCEPTIONS.member?(inst_id)
+      badge_id = idp == ht_institution.entityID ? :ok : :mismatch
```
Comment (aelkiss) [inline app/presenters/ht_registration_presenter.rb:112]: This one is definitely worth the "danger do not approve" label if they don't match

File: app/presenters/ht_registration_presenter.rb [hunk 5]
```diff
@@ -105,7 +122,29 @@ def show_detail_reverse_lookup
   end
 
   def show_detail_scoped_affiliation
-    env["HTTP_X_SHIB_EDUPERSONSCOPEDAFFILIATION"]
+    affiliation = env["HTTP_X_SHIB_EDUPERSONSCOPEDAFFILIATION"] || ""
+    # Check if scoped affiliation is faculty@, staff@, member@, or employee@
+    # AND matches the allowed affiliations for the IdP the user logged in with
+    badge = ""
+    unless self.class::AUTH_EXCEPTIONS.member? inst_id
+      badge_id = scoped_affiliation_valid? && scoped_affiliation_match? ? :ok : :mismatch
```
Comment (aelkiss) [inline app/presenters/ht_registration_presenter.rb:130]: Likewise I think the danger label is appropriate here

**Review by aelkiss — COMMENTED** [review]
I'll give it a try in dev. Tests look good; comments are primarily about how stringent a warning to display for certain kinds of mismatches.

File: app/presenters/ht_registration_presenter.rb (hunk 1 above)
Comment (moseshll) [reply app/presenters/ht_registration_presenter.rb:23]: Getting the affiliations regex into config requires YAML magic or `eval` and I'm not fired up about either of these. But the exceptions list seems to fit nicely.

Comment (aelkiss) [reply app/presenters/ht_registration_presenter.rb:23]: :+1: The valid affiliations seems less likely to change; if we find that we need to update them with any frequency we can see what we can do.

**Comment by moseshll** [comment]
Downgraded eppn and mail warnings to "questionable" label and non-red-alert tint.

File: app/presenters/ht_registration_presenter.rb [hunk 6]
```diff
@@ -105,7 +124,29 @@ def show_detail_reverse_lookup
...
+
+    affiliations = env["HTTP_X_SHIB_EDUPERSONSCOPEDAFFILIATION"].downcase.split(";")
+    affiliations.any? { |affiliation| self.class::VALID_AFFILIATIONS.match? affiliation }
+  end
+
+  def scoped_affiliation_match?
+    return false unless env["HTTP_X_SHIB_EDUPERSONSCOPEDAFFILIATION"].present?
+
+    affiliations = env["HTTP_X_SHIB_EDUPERSONSCOPEDAFFILIATION"].downcase.split(";")
+    affiliations.any? { |affiliation| ht_institution.allowed_affiliations.downcase.include? affiliation }
```
Comment (aelkiss) [inline app/presenters/ht_registration_presenter.rb:149]: I suspect this is why we're seeing a mismatch in testing -- `ht_institution.allowed_affiliations` is a regex

Comment (moseshll) [reply app/presenters/ht_registration_presenter.rb:149]: Curses, those pesky DB regexes.

File: test/presenters/ht_registration_presenter_test.rb [hunk 7]
```diff
@@ -89,3 +89,82 @@ def setup
     assert_match "/ht_registrations/#{reg.id}", reg.cancel_path
   end
 end
+
+class HTRegistrationPresenterENVTest < ActiveSupport::TestCase
+  def setup
+    @inst = create(:ht_institution, inst_id: "test",
+      allowed_affiliations: "faculty@default.invalid;staff@default.invalid")
```
Comment (aelkiss) [inline test/presenters/ht_registration_presenter_test.rb:96]: We would normally expect this to be `^(faculty|staff)@default.invalid`

Comment (moseshll) [reply test/presenters/ht_registration_presenter_test.rb]: Aaron Testuser Elkiss now shows OK badge in testing with latest patch.

**Review by aelkiss — APPROVED** [review]
Looks good on testing now. Merging.

**Pull request merged**